src/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -600,6 +611,155 @@ fromChunks cs FinalChunk{..} = CompactIndex {
     , ciLargerThanPage       = fcLargerThanPage
     }
 
+{-------------------------------------------------------------------------------
+  Serialisation
+-------------------------------------------------------------------------------}
+
+-- | 32 bit aligned.
+chunkBuilder :: Chunk -> BB.Builder
+chunkBuilder Chunk {..} = VU.foldMap BB.word32LE cPrimary
+
+-- | Must be written after the sequence of 'chunkBuilder' of the corresponding
+-- compact index. Specifically, the written chunks must match 'fcNumPages' to
+-- get the alignment right.
+finalChunkBuilder :: NumEntries -> FinalChunk -> BB.Builder
+finalChunkBuilder (NumEntries numEntries) FinalChunk {..} =
+       VU.foldMap BB.word32LE fcRangeFinder
+    <> (if odd (fcNumPages + VU.length fcRangeFinder)  -- align to 64 bit
+        then BB.word32LE 0
+        else mempty)
+    <> putBitVec fcClashes
+    <> putBitVec fcLargerThanPage
+    <> putTieBreaker fcTieBreaker
+    <> BB.word64LE (fromIntegral fcRangeFinderPrecision)
+    <> BB.word64LE (fromIntegral fcNumPages)
+    <> BB.word64LE (fromIntegral numEntries)
+
+-- | Padded to 64 bit.
+--
+-- Assumes that the bitvector is byte-aligned.
+putBitVec :: VU.Vector Bit -> BB.Builder
+putBitVec (BitVec offsetBits lenBits ba)
+  | mod8 offsetBits /= 0 = error "putBitVec: not byte aligned"
+  | otherwise =
+       -- first only write the bytes that are fully part of the bit vec
+       byteArrayFromTo offsetBytes offsetLastByte ba
+       -- then carefully write the last byte, might be partially uninitialised
```

Comment:
I was going to suggest that we could ensure the bitvec is initialised to all 0s, but I think on reflection what you've got is better, and it's clearly described.

Reply:
Oh and for the first byte we just assume the offset is byte-aligned (which is reasonable I think). Otherwise, we couldn't use `byteArrayFromTo` at all.

Reply:
> Oh and for the first byte we just assume the offset is byte-aligned (which is reasonable I think). Otherwise, we couldn't use byteArrayFromTo at all.

https://github.com/input-output-hk/lsm-tree/pull/108#discussion_r1512677382 is related, let's continue the discussion there